File: android/src/main/kotlin/com/shushikeji/improv_wifi_flutter/ImprovWifiFlutterPlugin.kt
package com.shushikeji.improv_wifi_flutter

import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothManager
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Build
import android.os.Handler
import android.os.Looper
import com.wifi.improv.DeviceState
import com.wifi.improv.ErrorState
import com.wifi.improv.ImprovDevice
import com.wifi.improv.ImprovManager
import com.wifi.improv.ImprovManagerCallback
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

class ImprovWifiFlutterPlugin :
    FlutterPlugin,
    MethodCallHandler,
    EventChannel.StreamHandler,
    ImprovManagerCallback {
    private lateinit var channel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private lateinit var applicationContext: Context
    private lateinit var improvManager: ImprovManager
    private lateinit var bluetoothManager: BluetoothManager
    private val discoveredDevices = mutableMapOf<String, ImprovDevice>()
    private val mainHandler = Handler(Looper.getMainLooper())
    private var eventSink: EventChannel.EventSink? = null
    private var bluetoothStateReceiver: BroadcastReceiver? = null

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        applicationContext = flutterPluginBinding.applicationContext
        bluetoothManager =
            applicationContext.getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager
        improvManager = ImprovManager(applicationContext, this)
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "improv_wifi_flutter")
        eventChannel =
            EventChannel(flutterPluginBinding.binaryMessenger, "improv_wifi_flutter/events")
        channel.setMethodCallHandler(this)
        eventChannel.setStreamHandler(this)
        registerBluetoothStateReceiver()
    }

    override fun onMethodCall(
        call: MethodCall,
        result: Result
    ) {
        when (call.method) {
            "getBluetoothState" -> result.success(getBluetoothStateName())
            "startScan" -> {
                if (getBluetoothStateName() != "poweredOn") {
                    emitEvent(
                        mapOf(
                            "type" to "scanFailed",
                            "reason" to "bluetoothUnavailable"
                        )
                    )
                    result.error(
                        "bluetooth_unavailable",
                        "Bluetooth must be powered on before scanning.",
                        null
                    )
                    return
                }
                improvManager.findDevices()
                result.success(null)
            }
            "stopScan" -> {
                improvManager.stopScan()
                result.success(null)
            }
            "connectToDevice" -> {
                val deviceId = call.argument<String>("deviceId")
                if (deviceId.isNullOrBlank()) {
                    result.error("invalid_argument", "deviceId is required.", null)
                    return
                }
                val device = discoveredDevices[deviceId]
                if (device == null) {
                    result.error(
                        "device_not_found",
                        "Device $deviceId must be discovered before connecting.",
                        null
                    )
                    return
                }
                improvManager.connectToDevice(device)
                result.success(null)
            }
            "identifyDevice" -> {
                try {
                    improvManager.identifyDevice()
                    result.success(null)
                } catch (exception: IllegalStateException) {
                    result.error("not_connected", exception.message, null)
                }
            }
            "sendWifi" -> {
                val ssid = call.argument<String>("ssid")
                val password = call.argument<String>("password")
                if (ssid == null || password == null) {
                    result.error("invalid_argument", "ssid and password are required.", null)
                    return
                }
                try {
                    improvManager.sendWifi(ssid, password)
                    result.success(null)
                } catch (exception: IllegalStateException) {
                    result.error("not_connected", exception.message, null)
                }
            }
            else -> result.notImplemented()
        }
    }

    override fun onListen(
        arguments: Any?,
        events: EventChannel.EventSink
    ) {
        eventSink = events
        emitEvent(
            mapOf(
                "type" to "bluetoothStateChanged",
                "bluetoothState" to getBluetoothStateName()
            )
        )
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    override fun onScanningStateChange(scanning: Boolean) {
        emitEvent(
            mapOf(
                "type" to "scanningStateChanged",
                "isScanning" to scanning
            )
        )
    }

    override fun onDeviceFound(device: ImprovDevice) {
        discoveredDevices[device.address] = device
        emitEvent(
            mapOf(
                "type" to "deviceFound",
                "device" to device.toMap()
            )
        )
    }

    override fun onConnectionStateChange(device: ImprovDevice?) {
        emitEvent(
            mapOf(
                "type" to "connectionStateChanged",
                "device" to device?.toMap()
            )
        )
    }

    override fun onStateChange(state: DeviceState) {
        emitEvent(
            mapOf(
                "type" to "deviceStateChanged",
                "deviceState" to state.toFlutterName()
            )
        )
    }

    override fun onErrorStateChange(errorState: ErrorState) {
        emitEvent(
            mapOf(
                "type" to "errorStateChanged",
                "errorState" to errorState.toFlutterName()
            )
        )
    }

    override fun onRpcResult(result: List<String>) {
        emitEvent(
            mapOf(
                "type" to "rpcResult",
                "rpcResult" to result
            )
        )
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        unregisterBluetoothStateReceiver()
    }

    private fun registerBluetoothStateReceiver() {
        val receiver = object : BroadcastReceiver() {
            override fun onReceive(
                context: Context?,
                intent: Intent?
            ) {
                if (intent?.action == BluetoothAdapter.ACTION_STATE_CHANGED) {
                    emitEvent(
                        mapOf(
                            "type" to "bluetoothStateChanged",
                            "bluetoothState" to getBluetoothStateName()
                        )
                    )
                }
            }
        }
        bluetoothStateReceiver = receiver

        val filter = IntentFilter(BluetoothAdapter.ACTION_STATE_CHANGED)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            applicationContext.registerReceiver(receiver, filter, Context.RECEIVER_NOT_EXPORTED)
        } else {
            @Suppress("DEPRECATION")
            applicationContext.registerReceiver(receiver, filter)
        }
    }

    private fun unregisterBluetoothStateReceiver() {
        bluetoothStateReceiver?.let {
            applicationContext.unregisterReceiver(it)
        }
        bluetoothStateReceiver = null
    }

    private fun getBluetoothStateName(): String {
        val adapter = bluetoothManager.adapter ?: return "unsupported"
        return when (adapter.state) {
            BluetoothAdapter.STATE_ON -> "poweredOn"
            BluetoothAdapter.STATE_OFF -> "poweredOff"
            BluetoothAdapter.STATE_TURNING_ON,
            BluetoothAdapter.STATE_TURNING_OFF -> "resetting"
            else -> "unknown"
        }
    }

    private fun emitEvent(event: Map<String, Any?>) {
        mainHandler.post {
            eventSink?.success(event)
        }
    }

    private fun ImprovDevice.toMap(): Map<String, Any?> {
        return mapOf(
            "id" to address,
            "name" to name
        )
    }

    private fun DeviceState.toFlutterName(): String {
        return when (this) {
            DeviceState.AUTHORIZATION_REQUIRED -> "authorizationRequired"
            DeviceState.AUTHORIZED -> "authorized"
            DeviceState.PROVISIONING -> "provisioning"
            DeviceState.PROVISIONED -> "provisioned"
        }
    }

    private fun ErrorState.toFlutterName(): String {
        return when (this) {
            ErrorState.NO_ERROR -> "noError"
            ErrorState.INVALID_RPC_PACKET -> "invalidRpcPacket"
            ErrorState.UNKNOWN_COMMAND -> "unknownCommand"
            ErrorState.UNABLE_TO_CONNECT -> "unableToConnect"
            ErrorState.NOT_AUTHORIZED -> "notAuthorized"
            ErrorState.UNKNOWN -> "unknown"
        }
    }
}

File: ios/Classes/ImprovWifiFlutterPlugin.swift
import Flutter
import CoreBluetooth

public class ImprovWifiFlutterPlugin: NSObject, FlutterPlugin, FlutterStreamHandler, ImprovManagerDelegate {
  private let improvManager: any ImprovManagerProtocol = ImprovManager.shared
  private var eventSink: FlutterEventSink?
  private var knownDeviceIds = Set<String>()

  public static func register(with registrar: FlutterPluginRegistrar) {
    let instance = ImprovWifiFlutterPlugin()
    let channel = FlutterMethodChannel(name: "improv_wifi_flutter", binaryMessenger: registrar.messenger())
    let eventChannel = FlutterEventChannel(name: "improv_wifi_flutter/events", binaryMessenger: registrar.messenger())

    registrar.addMethodCallDelegate(instance, channel: channel)
    eventChannel.setStreamHandler(instance)
  }

  override init() {
    super.init()
    improvManager.delegate = self
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "getBluetoothState":
      result(mapBluetoothState(improvManager.bluetoothState))
    case "startScan":
      improvManager.reset()
      improvManager.scan()
      result(nil)
    case "stopScan":
      improvManager.stopScan()
      result(nil)
    case "connectToDevice":
      guard let args = call.arguments as? [String: Any],
            let deviceId = args["deviceId"] as? String,
            let peripheral = improvManager.foundDevices[deviceId] else {
        result(FlutterError(code: "device_not_found", message: "Device must be discovered before connecting.", details: nil))
        return
      }
      improvManager.connectToDevice(peripheral)
      result(nil)
    case "identifyDevice":
      if let error = improvManager.identifyDevice() {
        result(flutterError(from: error))
      } else {
        result(nil)
      }
    case "sendWifi":
      guard let args = call.arguments as? [String: Any],
            let ssid = args["ssid"] as? String,
            let password = args["password"] as? String else {
        result(FlutterError(code: "invalid_argument", message: "ssid and password are required.", details: nil))
        return
      }

      if let error = improvManager.sendWifi(ssid: ssid, password: password) {
        result(flutterError(from: error))
      } else {
        result(nil)
      }
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    eventSink = events
    emitEvent(
      type: "bluetoothStateChanged",
      extra: ["bluetoothState": mapBluetoothState(improvManager.bluetoothState)]
    )
    emitEvent(
      type: "scanningStateChanged",
      extra: ["isScanning": improvManager.scanInProgress]
    )
    return nil
  }

  public func onCancel(withArguments arguments: Any?) -> FlutterError? {
    eventSink = nil
    return nil
  }

  public func didUpdateBluetoohState(_ state: CBManagerState) {
    emitEvent(
      type: "bluetoothStateChanged",
      extra: ["bluetoothState": mapBluetoothState(state)]
    )
  }

  public func didUpdateFoundDevices(devices: [String : CBPeripheral]) {
    for (deviceId, peripheral) in devices where !knownDeviceIds.contains(deviceId) {
      emitEvent(type: "deviceFound", extra: ["device": deviceMap(from: peripheral)])
    }
    knownDeviceIds = Set(devices.keys)
  }

  public func didConnect(peripheral: CBPeripheral) {
    if peripheral.state == .connected {
      emitEvent(
        type: "connectionStateChanged",
        extra: ["device": deviceMap(from: peripheral)]
      )
    } else {
      emitEvent(type: "connectionStateChanged")
    }
  }

  public func didDisconnect(peripheral _: CBPeripheral) {
    emitEvent(type: "connectionStateChanged")
  }

  public func didUpdateDeviceState(_ state: DeviceState?) {
    var extra = [String: Any]()
    if let stateName = mapDeviceState(state) {
      extra["deviceState"] = stateName
    }
    emitEvent(type: "deviceStateChanged", extra: extra)
  }

  public func didUpdateErrorState(_ state: ErrorState?) {
    var extra = [String: Any]()
    if let stateName = mapErrorState(state) {
      extra["errorState"] = stateName
    }
    emitEvent(type: "errorStateChanged", extra: extra)
  }

  public func didReceiveResult(_ result: [String]?) {
    emitEvent(type: "rpcResult", extra: ["rpcResult": result ?? []])
  }

  public func didReset() {
    knownDeviceIds.removeAll()
    emitEvent(type: "connectionStateChanged")
    emitEvent(type: "deviceStateChanged")
    emitEvent(type: "errorStateChanged")
    emitEvent(type: "rpcResult", extra: ["rpcResult": []])
    emitEvent(type: "scanningStateChanged", extra: ["isScanning": false])
  }

  public func didUpdateIsScanning(_ isScanning: Bool) {
    emitEvent(type: "scanningStateChanged", extra: ["isScanning": isScanning])
  }

  public func didFailScanningBluetoothNotAvailable() {
    emitEvent(type: "scanFailed", extra: ["reason": "bluetoothUnavailable"])
  }

  private func emitEvent(type: String, extra: [String: Any] = [:]) {
    var payload: [String: Any] = ["type": type]
    for (key, value) in extra {
      payload[key] = value
    }
    eventSink?(payload)
  }

  private func deviceMap(from peripheral: CBPeripheral) -> [String: Any] {
    return [
      "id": peripheral.identifier.uuidString,
      "name": peripheral.name ?? NSNull(),
    ]
  }

  private func mapBluetoothState(_ state: CBManagerState) -> String {
    switch state {
    case .unknown:
      return "unknown"
    case .resetting:
      return "resetting"
    case .unsupported:
      return "unsupported"
    case .unauthorized:
      return "unauthorized"
    case .poweredOff:
      return "poweredOff"
    case .poweredOn:
      return "poweredOn"
    @unknown default:
      return "unknown"
    }
  }

  private func mapDeviceState(_ state: DeviceState?) -> String? {
    switch state {
    case .authorizationRequired:
      return "authorizationRequired"
    case .authorized:
      return "authorized"
    case .provisioning:
      return "provisioning"
    case .provisioned:
      return "provisioned"
    case .none:
      return nil
    }
  }

  private func mapErrorState(_ state: ErrorState?) -> String? {
    switch state {
    case .noError:
      return "noError"
    case .invalidRPCPacket:
      return "invalidRpcPacket"
    case .unknownCommand:
      return "unknownCommand"
    case .unableToConnect:
      return "unableToConnect"
    case .notAuthorized:
      return "notAuthorized"
    case .unknown:
      return "unknown"
    case .none:
      return nil
    }
  }

  private func flutterError(from error: BluetoothManagerError) -> FlutterError {
    switch error {
    case .deviceDisconnected:
      return FlutterError(code: "not_connected", message: "No device is currently connected.", details: nil)
    case .serviceNotAvailable:
      return FlutterError(code: "service_not_available", message: "Improv service is not available on the connected device.", details: nil)
    }
  }
}
